#3 Signal inputs

devversion opened the pull request:
See individual commits and #53521 for more context.

devversion committed:
test(core): add test playground for zone signal inputs

This commit creates a small http server Angular application playground
for playing with signal inputs. This is useful for development and also
validates some of the common input patterns.

devversion committed:
refactor(compiler-cli): improve type safety of linker AST

The linker AST is abstracted to be agnostic of the underlying
implementation AST. i.e. TS AST or Babel AST.

This abstraction also intends to provide some type-safety-ness to
parsing of various partial declarations. This commit improves type
safety further by fixing that `AstValue'`s were not checked for
assignability of `T`- potentially hiding issues/unaccounted values.

Additionally, we fix that `getObject()` does not properly narrow
union types to actual object literals. This happend because e.g. arrays
are of type `object`. We can improve type safety here. Using `Record`
did not help as an array would still assign to that.

devversion committed:
build: fix partial output generation errors swallowed

In some cases, the input files for a partial output generation
compliance tests may be invalid and lead to compilation errors.

The golden partial would be silently generated with the remaining
test cases. Instead of hiding errors, we will now print these and
cause the script to fail properly.

Note that the error logging is pretty minimalistic, but it's sufficient.

devversion committed:
refactor(compiler-cli): support linking/parsing new partial output for signal inputs

As we introduced the new partial output for signal inputs, we also need
to update the linker code to be able to parse this. This commit adds
this functionality.

In the follow-up commit, compliance tests for linking, partial output,
and full compilations are added.

devversion committed:
test: add signal input compliance tests for linking, partial output and full compilations

Adds signal input compliance tests, ensuring linking works as expected,
partial output is generated properly, types are inferred properly, and
that the full, or linked output matches our expected runtime structure.

devversion committed:
test: add integration test for signal inputs in CLI

This commit introduces a new integration test to ensure signal inputs
work as expected for end uses in Angular CLI applications.

devversion committed:
refactor(core): improve error message and add guide for required inputs

Whenever a required input is accessed too early in a
directive/component, the signal input will throw an error.

This is necessary so that we can support required inputs
with intuitive typings that do not include `undefined` for
the short period of time where Angular is creating the component and
then assigning inputs later (Angular currently has no way of setting
inputs as part of the class creation when `new Dir()` happens)

devversion committed:
test(core): add runtime tests for input signal

This commit adds some runtime unit tests to ensure that input
signal is behaving properly at runtime.

devversion committed:
test(compiler-cli): add additional diagnostic test for generic inline constructors

We recently landed a commit to introduce support for generic type
checking of signal inputs. For that we had to implement logic that
will generate imports for inline type constructors. This required
changes to the context logic and `TypeCtorOp` file-level op.

This commit ensures that everything is working as expected, specifically
in cases where an inline type ctor is generated and imports would be
needed to unwrap the class members for `InputSignal`.

devversion committed:
refactor(compiler-cli): follow-up typo fixes for typecheck utilities

Follow-up for the inital signal inputs PR, fixing a typo that was made.

devversion committed:
refactor(core): add internal signal input support for `@Input` decorator

We are adding internal support for declaring signal inputs via the
`@Input` decorator. This is needed for JIT unit testing, or JIT
applications.

In JIT, Angular is not able to recognize signal inputs due to the
lack of static reflection metadata. Decorators attach their information
on the class- without it needing to be instantiated. This allows Angular
to know inputs when preparing/generating the directive definition. With
signal inputs this is not possible- so we need a way to tell Angular
about inputs for JIT applications. We've decided that this is not
something users should have to deal with, so a transform will be added
in a follow-up that will automatically derive/and add the decorators
for signal inputs when requested in JIT environments.

devversion committed:
refactor(compiler-cli): expose helper for parsing signal input class members

Moves the signal input class member extraction logic into the dedicated
input function file. This makes the code for signal inputs more
self-contained.

This commit then re-exposes the function as part of `ngtsc/annotations`
so that it can be used later for a transform that will take the signal
input metadata and translate it into a `@Input` decorator. This allows
us to remove code duplication and guarantees consistency/correctness

devversion committed:
refactor(compiler-cli): expose import manager import insertion helper

This commit does two things:

- exposes `addImports` so that it can be used by transforms that we are
  adding to the compiler. e.g. the signal input to `@Input` transform.

- `updates `addImports` to support/use the transform context factory.
  This will allow us to write proper transforms using `addImports`. Also
  leverages this in the Ivy JS/DTS transforms.

devversion committed:
refactor(compiler-cli): add transform to support signal input in JIT

This commit adds a transform for supporting input signals in JIT
environments. The transform will be wired up for Angular CLI
applications automatically. An integration test verifies that this fixes
unit testing with signal inputs.

The transform basically will take the signal input metadata and
transform it into `@Input` decorators that can provide static
information to the Angular JIT runtime when the directive/component
definition is compiled.

eneajaho commented on packages/core/src/metadata/directives.ts:
```diff
@@ -832,6 +832,20 @@ export interface Input {
+   *
+   * This option exists for JIT compatibility. Users are not expected to use this.
+   * Angular needs a way to capture inputs from classes so that the internal data
+   * structures can be et up. This needs to happen before the component is instantiated.
```
```suggestion
   * structures can be set up. This needs to happen before the component is instantiated.
```

devversion committed:
test: add completion tests for language service with signal inputs

Given that the TCB output changes with signal inputs, and one of our
important considerations was auto-completion, we need some unit tests
that verify and guarantee proper completion with signal inputs being
bound in templates. This commit adds these.

devversion committed:
refactor(compiler-cli): diagnostics ensuring signal inputs are declared properly

This commit addds two diagnostics for two scenarios where signal inputs
are declared incorrectly:

- a signal input is also annotated with `@Input` in the TypeScript
  sources.
- a signal input is also declared in the `inputs` option of
  `@Directive/`@Component`.

devversion committed:
refactor(compiler-cli): add diagnostic for inputs declared on static members

Currently when someone declares a signal or non-signal input on a static
class member, the compiler will not yield any diagnostic. We can detect
these mistakes and report a diagnostic to help our users.

devversion committed:
refactor(compiler-cli): drop dependency on external modules for input recognition

Improves the recognition of the `input`/`input.required` functions to
not depend on external module resolution. This is useful for local
compilation and for transforms operating on a single file/ isolated
module.

devversion committed:
test: add infrastructure to run signal acceptance tests with JIT and AOT

Adds infrastructure to run signal input tests with JIT (using the
transform) and AOT. Acceptance tests for signal inputs will run with
both variants. In the future we can consider expanding this
infrastructure for all of our acceptance tests, but that's a different
story.

devversion committed:
refactor(core): properly emit input flags using bitwise or

As part of testing we did accidentally use `bitwiseAnd` for the input
flags, given we started without an extra flag for `HasTransform`.

This commit teaches the compiler to support emitting bitwise OR
and uses it when combining input flags, fully re-enabling transforms
for signal components after the new flag mechanism was introduced in
previous commits.

devversion committed:
test(core): add acceptance unit tests for signal inputs

Adds AOT and JIT unit tests for signal inputs that verify integration of
signal inputs for our users.

JoostK reviewed (commented):
Left mostly nits

JeanMeche commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -106,10 +107,13 @@ function toInputMapping<TExpression>(
         'Unsupported input, expected a string or an array containing two strings and an optional function');
   }
 
+  const transformFunction = values.length > 2 ? values[2].getOpaque() : null;
```
You need a length of `3` to access index `2` or a I missing something ?

JoostK replied:
> You need a length of `3` to access index `2` or a I missing something ? 

No you are correct. That's what happens with midnight reviews 🥹

pkozlowski-opensource commented on adev/src/content/reference/errors/NG0950.md:
```diff
@@ -0,0 +1,15 @@
+This can happen when a required input is accessed too early in your directive or component.
+This is commonly happening when the input is read as part of class construction.
+
+Angular will set required inputs at earliest in the `ngOnInit` lifecycle hook.
```
+1, agree with @JoostK here

pkozlowski-opensource commented on adev/src/content/reference/errors/NG0950.md:
```diff
@@ -0,0 +1,15 @@
+
+## Fixing the error
+
+Access the required input in `ngOnInit` or later.
```
IMO the no 1 recommendation would be to read inputs in the reactive context (computed, effect, expression in a template).

devversion replied:
Agreed, mentioning this a bit, but we need much better documentation on reactive contexts at some point. Something I can link to.

pkozlowski-opensource reviewed (approved):
LGTM! I went through the runtime / core parts and only skimmed through the compiler / integration tests bits.

devversion commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -84,18 +84,38 @@ function toInputMapping<TExpression>(
+      isSignal: obj.getBoolean('isSignal'),
+      required: obj.getBoolean('isRequired'),
+      hasTransform: obj.getBoolean('hasTransform'),
+      transformFunction: transformValue.isNull() ? null : transformValue.getOpaque(),
```
Yeah, agreed. I went with explicitness as partial output is not optimized for size and this feels more clear

devversion commented on packages/compiler-cli/private/tooling.ts:
```diff
@@ -33,17 +34,48 @@ export const GLOBAL_DEFS_FOR_TERSER_WITH_AOT = {
+  return (ctx) => {
+    return (sourceFile) => {
+      sourceFile = downlevelDecoratorTransform(ctx)(sourceFile);
+      sourceFile = inputSignalMetadataTransform(ctx)(sourceFile);
```
Not intentional at all, good point. In a mixed situation, reflection capabilities does indeed only look at either field, so that is a good point. Thx

devversion commented on packages/compiler-cli/test/ngtsc/authoring_spec.ts:
```diff
@@ -36,6 +36,40 @@ runInEachFileSystem(() => {
+      })]);
+    });
+
+    it('should fail if signal input is also declared in `inputs` decorator field.', () => {
```
Those were different issues (with existing non-signal inputs), but will add one here. Thanks!

devversion committed:
refactor(core): detect signal input transforms independently of flag

This commit changes the `HasTransform` flag to be only concerned with
decorator inputs. This allows us to automatically detect signal input
transforms without reliance on the flag, resulting in less complexity in
the compiler (as outlined in the design doc) and various  other places,
while it also allows us to simplify JIT support for signal inputs
because there would be no need to capture the "hasTransform" state in
the decorator so that JIT can generate the according input flags.

`isSignal` will still persist as an input flag to allow for monomorphic
and highly efficient distinguishing at runtime, whether an input is
signal based or not. JIT transform will also need to propagate this
information to the runtime somehow.

alan-agius4 commented on integration/cli-signal-inputs/angular.json:
```diff
@@ -0,0 +1,119 @@
+{
```
Would be nice to use the new esbuild based system and also remove usages of deprecated options.

```json
{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "cli-signal-inputs": {
      "projectType": "application",
      "schematics": {},
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "outputPath": "dist",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "polyfills": ["zone.js"],
            "tsConfig": "tsconfig.app.json",
            "aot": true,
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": ["src/styles.css"],
            "scripts": [],
            "optimization": false,
            "progress": false
          },
          "configurations": {
            "production": {
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "namedChunks": false,
              "extractLicenses": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "options": {
            "buildTarget": "cli-signal-inputs:build"
          },
          "configurations": {
            "dev": {
              "buildTarget": "cli-signal-inputs:build:dev"
            },
            "production": {
              "buildTarget": "cli-signal-inputs:build:production"
            }
          }
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "buildTarget": "cli-signal-inputs:build"
          }
        },
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "polyfills": ["zone.js", "zone.js/testing"],
            "tsConfig": "tsconfig.spec.json",
            "karmaConfig": "karma.conf.js",
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": ["src/styles.css"],
            "scripts": [],
            "progress": false,
            "watch": false
          }
        },
        "e2e": {
          "builder": "@angular-devkit/build-angular:protractor",
          "options": {
            "protractorConfig": "e2e/protractor.conf.js",
            "devServerTarget": "cli-signal-inputs:serve",
            "webdriverUpdate": false
          },
          "configurations": {
            "production": {
              "devServerTarget": "cli-signal-inputs:serve:production"
            }
          }
        }
      }
    }
  }
}
```

devversion replied:
Agreed. done. Looks like most integration tests would need this (I just copied one boilerplate)

alan-agius4 reviewed (approved):
LGTM, one nit.

Reviewed-for: tooling-cli-shared-api

devversion committed:
fixup! test: add integration test for signal inputs in CLI

Update integration test CLI app

devversion commented:
Caretaker note: pre-existing failures. TGP is overall green, unaffected targets

devversion commented:
This PR was merged into the repository by commit 7862686a600d043f3b3ebeafb23ee4a29647d3bf.
